feat: monkey-patch MessagePort onproperties

JiaLiPassion opened the pull request:
Close #34581

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #34581 


## What is the new behavior?
monkey-patch MessagePort.onmessage/onmessageerror properties. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

alfaproject commented on packages/zone.js/test/browser/messageport.spec.ts:
```diff
@@ -0,0 +1,45 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+fdescribe('MessagePort', () => {
```
eff

JiaLiPassion replied:
thanks, forget to remove.

mhevery commented on packages/zone.js/lib/browser/message-port.ts:
```diff
@@ -0,0 +1,13 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+Zone.__load_patch('MessagePort', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
```
please leave some space as it looks like the comment belongs to the function call.

JiaLiPassion replied:
Got it, thanks, updated.

mhevery commented on packages/zone.js/test/browser/messageport.spec.ts:
```diff
@@ -0,0 +1,45 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+describe('MessagePort', () => {
```
please leave some space as it looks like the comment belongs to the function call.

JiaLiPassion replied:
Got it, updated , thanks.

mhevery reviewed (approved):
Approved with some nits. 

Also the tests are failing, usually we try to get CI green before doing code review.

JiaLiPassion commented:
@mhevery , thanks, the test failed because there is a flaky cases, I fixed it here, https://github.com/angular/angular/pull/35679

mhevery commented:
[presubmit](http://test/OCL:299418820:BASE:299418851:1583528752674:b6118229)

AndrewKushnir commented:
FYI, the changes from this PR caused some failures in g3, so the changes were reverted from master. I'm re-opening this ticket.

cc @matsko

JiaLiPassion commented:
@AndrewKushnir , could you tell me more information about the failure in g3, because this PR should have no impact for the current code, this PR just build a new `bundle` , and this bundle will not be loaded by default. Thanks!

AndrewKushnir commented:
Hi @JiaLiPassion,

If I understand correctly, all files in the `browser` folder are included (except excluded ones), here is a snippet from the Angular BUILD file in g3:

```
ts_library(
    name = "zonejs_lib",
    srcs = glob(
        [
            "packages/zone.js/lib/browser/*.ts",
        ],
        exclude = [
            "packages/zone.js/lib/browser/rollup-*.ts",
            "packages/zone.js/lib/browser/shadydom.ts",
            "packages/zone.js/lib/browser/utils.ts",
            "packages/zone.js/lib/browser/webapis-*.ts",
        ],
    ) + [
        "rc/packages/zone.js/lib/zone-spec/long-stack-trace.ts",
        "rc/packages/zone.js/lib/zone-spec/wtf.ts",
    ],
    ...
)
```
So the newly added `packages/zone.js/lib/browser/message-port.ts` was also included and as a result affected some code.

Should it be in the `exclude` list?

Thank you.

JiaLiPassion commented:
@AndrewKushnir , oh, yes, I see the code snippet also exclude `webapi-*` so this `messageport` should also be excluded. This is an optional packages, so I think `g3` doesn't need this file to be compiled. 
Thanks.

AndrewKushnir commented:
Hi @JiaLiPassion, thanks for the update. Could you please rebase this PR, so that we re-run CI again and I'll try to run presubmit (excluding the file) in g3 again. Thank you.

JiaLiPassion committed:
feat(zone.js): Monkey patch MessagePort.prototype onproperties

Monkey patch `MessagePort.prototype.onmessage` and `MessagePort.prototype.onmessageerror` to make
these properties's value(callback function) run in the zone when these value are set.

JiaLiPassion commented:
@AndrewKushnir , sure, I have rebased this PR, please review, thanks!

AndrewKushnir commented:
@JiaLiPassion thanks for the update.

FYI, in order to run tests for ZoneJS changes that are in the merge queue, I've combined 4 ZoneJS PRs into one: https://github.com/angular/angular/pull/36043. Note: the https://github.com/angular/angular/pull/36043 PR is for testing purposes only and the changes will land in the corresponding PRs once g3 testing is completed.

Thank you.

JiaLiPassion commented:
@AndrewKushnir, got it, thank you!

AndrewKushnir commented:
FYI, changes in this PR were tested against g3 as a part of https://github.com/angular/angular/pull/36043, using Global TAP run, [which was successful](https://test.corp.google.com/ui#id=OCL:300603599:BASE:301077666:1584325797358:d8800cfc&include_skipped=true). Thank you.
